### PR TITLE
[dvsim] Add GUI mode for running simulations

### PR DESCRIPTION
**Pull request by sriyerg**
The adds support for running simulations in GUI mode. This change plumbs
the dvsim switch `--gui` to the underlying tools. With VCS and Xcelium,
the respective GUI windows will open, exposing the UCLI prompt, where
the user can take control of running the simulation (debugging, adding
breakpoints etc).

If GUI mode is enabled and multiple tests are provided for run, it picks
the first and drops everything else. The onus is on the user to pick
correctly (pass a single test with `--items` and a specific seed with
`--seed`).

Further, in GUI mode, it drops the pass and fail patterns, since the
whole simulation is run from inside the tool (the log file is not
generated).

Only VCS and Xcelium are currently fully supported. For all others,
`--gui` has no effect.

Signed-off-by: Srikrishna Iyer <sriyer@google.com>

### CODE REVIEW

File: hw/dv/tools/dvsim/verilator.hjson [hunk 1]
```diff
@@ -108,6 +108,12 @@
   cov_db_test_dir: ""
 
   build_modes: [
+    {
+      name: verilator_gui
```
Comment (rswarbrick) [inline hw/dv/tools/dvsim/verilator.hjson:112]: This seems a bit silly. Could we do it the other way, where we check whether `$tool_gui` exists as a build mode, and only use it if so?

File: hw/dv/tools/dvsim/xcelium.hjson [hunk 2]
```diff
@@ -5,7 +5,7 @@
   build_cmd:  "{job_prefix} xrun"
   run_cmd:    "{job_prefix} xrun"
 
-  build_opts: ["-elaborate -64bit -access +r -sv",
+  build_opts: ["-elaborate -64bit -sv",
```
Comment (rswarbrick) [inline hw/dv/tools/dvsim/xcelium.hjson:8]: Hmm, could you split this and the xcelium_waves change into a separate commit and write a commit message explaining why it's correct? I'm just thinking that someone (maybe me!) is going to want to do archaeology on the history of `xcelium.hjson` sometime in the future.

File: util/dvsim/SimCfg.py [hunk 3]
```diff
@@ -504,6 +507,12 @@ def _create_deploy_objects(self):
         self.runs = ([]
                      if self.build_only else self._expand_run_list(build_map))
 
+        # In GUI mode, only allow one test to run.
+        if self.gui and len(self.runs) > 1:
+            self.runs = self.runs[:1]
+            log.error("In GUI mode, only one test is allowed to run. "
+                      "Picking {}".format(self.runs[0].full_name))
```
Comment (rswarbrick) [inline util/dvsim/SimCfg.py:514]: I think this should either be `log.error` and exit immediately with a failure code or `log.warning`.

File: util/dvsim/dvsim.py [hunk 4]
```diff
@@ -402,6 +402,12 @@ def parse_args():
                         help=('The options for each build_mode in this list '
                               'are applied to all build and run targets.'))
 
+    disg.add_argument("--gui",
+                      action='store_true',
+                      default=False,
```
Comment (rswarbrick) [inline util/dvsim/dvsim.py:407]: Nit: `store_true` already sets the default value, so you don't actually need that here.

File: util/dvsim/dvsim.py [hunk 5]
```diff
@@ -660,19 +666,19 @@ def main():
...
-        sys.exit(0)
-
-    # In simulation mode: if --cov-analyze switch is passed, then run the GUI
-    # tool.
-    if args.cov_analyze:
-        cfg.cov_analyze()
-        cfg.deploy_objects()
-        sys.exit(0)
+    # Run sub-flows (based on switches passed).
+    if cfg.flow == 'sim':
```
Comment (rswarbrick) [inline util/dvsim/dvsim.py:670]: Is this change related to GUI support? If not, maybe it's clearer to split it out into a separate commit?

File: hw/dv/tools/dvsim/xcelium.hjson (hunk 2 above)
Comment (sriyerg) [reply hw/dv/tools/dvsim/xcelium.hjson:8]: The removal of this switch (duplicated on line 23) has no effect on Xcelium's behavior.

File: hw/dv/tools/dvsim/verilator.hjson (hunk 1 above)
Comment (sriyerg) [reply hw/dv/tools/dvsim/verilator.hjson:114]: If a mode is enabled but it doesn't exist, dvsim throws an error and exits (which is the correct behavior). How to deal with what you are asking for probably does not belong with this PR. Added a TODO to figure it out later.

File: util/dvsim/SimCfg.py (hunk 3 above)
Comment (sriyerg) [reply util/dvsim/SimCfg.py]: Updated.

File: util/dvsim/dvsim.py (hunk 4 above)
Comment (sriyerg) [reply util/dvsim/dvsim.py]: Updated.

File: util/dvsim/dvsim.py (hunk 5 above)
Comment (sriyerg) [reply util/dvsim/dvsim.py]: This is not related to GUI change but it was just cosmetic. Reverted anyway because it does not matter.

File: hw/dv/tools/dvsim/verilator.hjson (hunk 1 above)
Comment (rswarbrick) [reply hw/dv/tools/dvsim/verilator.hjson:114]: Understood. Presumably, the error currently looks like some weird internal "aargh, something went wrong" message. It would be nice (as a future change) to print out something like "This tool doesn't have a GUI" to make it clear to the user what's going on.

File: hw/dv/tools/dvsim/xcelium.hjson (hunk 2 above)
Comment (rswarbrick) [reply hw/dv/tools/dvsim/xcelium.hjson:8]: Oh, I see! I hadn't realised that it was duplicated. Ok, cool.

**Review by rswarbrick — APPROVED** [review]

**Review by weicaiyang — APPROVED** [review]
LGTM, thanks for adding this.
